types/indexer_block_events.go

```diff
@@ -0,0 +1,93 @@
+package types
+
+import (
+	"time"
+)
+
+type IndexerBlockEventManager struct {
+	height   uint32
+	time     time.Time
+	txHashes []string
+	// map from tx hash to list of tx events
+	txEventsMap map[string][]*IndexerTendermintEvent
+}
+
+// NewIndexerBlockEventManager returns a new IndexerBlockEventManager.
+// This should be called in BeginBlocker.
+func NewIndexerBlockEventManager(height uint32, time time.Time) *IndexerBlockEventManager {
+	return &IndexerBlockEventManager{
+		txHashes:    []string{},
+		txEventsMap: make(map[string][]*IndexerTendermintEvent),
+		height:      height,
+		time:        time,
+	}
+}
+
+// MergeEvents merges the events from another block event manager into this one if
+// the block heights are the same.
+func (eventManager *IndexerBlockEventManager) MergeEvents(mgr *IndexerBlockEventManager) {
+	if eventManager.height != mgr.height {
+		return
+	}
+	for _, txHash := range mgr.txHashes {
+		events := mgr.txEventsMap[txHash]
+		for _, event := range events {
+			eventManager.AddTxnEvent(txHash, event.Subtype, event.Data)
+		}
+	}
+}
+
+// AddTxnEvent adds a transaction event to the block event manager. If the transaction hash is not already in the
+// block event manager, it is also added.
+func (eventManager *IndexerBlockEventManager) AddTxnEvent(txHash string, subType string, data string) {
```

Comment:
nit: Do we need a TODO for a method to add block events?

Reply:
done